cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java

```diff
@@ -0,0 +1,439 @@
+import com.google.common.base.Joiner;
+import com.google.common.base.Splitter;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.google.common.io.Closeables;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonToken;
+import com.google.gson.stream.JsonWriter;
+import it.unimi.dsi.fastutil.bytes.ByteArrayList;
+import it.unimi.dsi.fastutil.bytes.ByteList;
+
+import java.io.IOException;
+import java.io.StringReader;
+import java.io.StringWriter;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Utility class for converting {@link StructuredRecord} to and from json.
+ */
+public final class StructuredRecordStringConversion {
+
+  // Known Java type to schema type mapping
+  // Doesn't have map and array as those need to use instanceof to check
+  private static final Map<Class<?>, Schema.Type> TYPE_TO_SCHEMA = new IdentityHashMap<Class<?>, Schema.Type>(
+    ImmutableMap.<Class<?>, Schema.Type>builder()
+      .put(Boolean.class, Schema.Type.BOOLEAN)
+      .put(Byte.class, Schema.Type.INT)
+      .put(Short.class, Schema.Type.INT)
+      .put(Integer.class, Schema.Type.INT)
+      .put(Long.class, Schema.Type.LONG)
+      .put(Float.class, Schema.Type.FLOAT)
+      .put(Double.class, Schema.Type.DOUBLE)
+      .put(String.class, Schema.Type.STRING)
+      .put(ByteBuffer.class, Schema.Type.BYTES)
+      .put(byte[].class, Schema.Type.BYTES)
+      .put(StructuredRecord.class, Schema.Type.RECORD)
+      .build()
+  );
+
+  private static final EnumMap<Schema.Type, JsonToken> SCHEMA_TO_JSON_TYPE = new EnumMap<Schema.Type, JsonToken>(
+    ImmutableMap.<Schema.Type, JsonToken>builder()
+      .put(Schema.Type.NULL, JsonToken.NULL)
+      .put(Schema.Type.BOOLEAN, JsonToken.BOOLEAN)
+      .put(Schema.Type.INT, JsonToken.NUMBER)
+      .put(Schema.Type.LONG, JsonToken.NUMBER)
+      .put(Schema.Type.FLOAT, JsonToken.NUMBER)
+      .put(Schema.Type.DOUBLE, JsonToken.NUMBER)
+      .put(Schema.Type.STRING, JsonToken.STRING)
+      .put(Schema.Type.BYTES, JsonToken.BEGIN_ARRAY)
+      .put(Schema.Type.ARRAY, JsonToken.BEGIN_ARRAY)
+      .put(Schema.Type.MAP, JsonToken.BEGIN_OBJECT)
+      .put(Schema.Type.RECORD, JsonToken.BEGIN_OBJECT)
+      .build()
+  );
+
+  /**
+   * Converts a {@link StructuredRecord} to a json string.
+   */
+  public static String toJsonString(StructuredRecord record) {
+    StringWriter strWriter = new StringWriter();
+    JsonWriter writer = new JsonWriter(strWriter);
+    try {
+      writeJson(writer, record.getSchema(), record);
+      return strWriter.toString();
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
```

Comment:
should we throw checked exceptions here, and leave it to the caller to handle them?